servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/PublisherProcessorBuffers.java

```diff
@@ -0,0 +1,88 @@
+     * consumed. If more items are {@link PublisherProcessorBuffer#add(Object) added} to the returned
+     * {@link PublisherProcessorBuffer} then that {@link PublisherProcessorBuffer#add(Object) addition} will fail.
+     *
+     * @param maxBuffer Maximum number of items that can be present in the returned
+     * @param <T> Type of items added to the returned {@link PublisherProcessorBuffer}.
+     * @return A new {@link PublisherProcessorBuffer}.
+     */
+    public static <T> PublisherProcessorBuffer<T> fixedSize(final int maxBuffer) {
+        return new AbstractPublisherProcessorBuffer<T, Queue<Object>>(maxBuffer, newMpscQueue(2, maxBuffer)) {
+            @Override
+            void offerPastBufferSize(final Object signal, final Queue<Object> queue) {
+                throw new QueueFullException("processor-buffer", maxBuffer);
+            }
+        };
+    }
+
+    /**
+     * Creates a new {@link PublisherProcessorBuffer} which buffers a maximum of {@code maxBuffer} items without being
+     * consumed. If more items are {@link PublisherProcessorBuffer#add(Object) added} to the returned
+     * {@link PublisherProcessorBuffer} then that {@link PublisherProcessorBuffer#add(Object) addition} will be dropped.
+     *
+     * @param maxBuffer Maximum number of items that can be present in the returned
+     * @param <T> Type of items added to the returned {@link PublisherProcessorBuffer}.
+     * @return A new {@link PublisherProcessorBuffer}.
+     */
+    public static <T> PublisherProcessorBuffer<T> fixedSizeDropLatest(final int maxBuffer) {
+        return new AbstractPublisherProcessorBuffer<T, Queue<Object>>(maxBuffer, newMpscQueue(2, maxBuffer)) {
+            @Override
+            void offerPastBufferSize(final Object signal, final Queue<Object> queue) {
+                // noop => drop latest
+            }
+        };
+    }
+
+    /**
+     * Creates a new {@link PublisherProcessorBuffer} which buffers a maximum of {@code maxBuffer} items without being
+     * consumed. If more items are {@link PublisherProcessorBuffer#add(Object) added} to the returned
+     * {@link PublisherProcessorBuffer} then the oldest item previously added to the buffer will be dropped.
+     *
+     * @param maxBuffer Maximum number of items that can be present in the returned
+     * @param <T> Type of items added to the returned {@link PublisherProcessorBuffer}.
+     * @return A new {@link PublisherProcessorBuffer}.
+     */
+    public static <T> PublisherProcessorBuffer<T> fixedSizeDropOldest(final int maxBuffer) {
+        return new AbstractPublisherProcessorBuffer<T, ConcurrentLinkedQueue<Object>>(maxBuffer,
+                new ConcurrentLinkedQueue<>()) {
+            @Override
+            void offerPastBufferSize(final Object signal, final ConcurrentLinkedQueue<Object> queue) {
+                queue.poll(); // drop oldest
+                queue.offer(signal);
```

Comment:
I think this is racy as you may have `queue.offer` return false if this method is invoked by multiple threads.

You should do something like:
```java
void offerPastBufferSize(final Object signal, final ConcurrentLinkedQueue<Object> queue) {
    while (!queue.offer(signal)) {
        queue.poll();
    }
}
```

Reply:
still I think it should be done to guard against problems later when the code is changed. At least I would add a comment with a warning that the queue must be unbound.